diff --git a/CMakeLists.txt b/CMakeLists.txt
index c24c79f..3bd6699 100644
--- a/CMakeLists.txt
+++ b/CMakeLists.txt
@@ -95,58 +95,3 @@ if(BACKWARD_HAS_EXTERNAL_LIBRARIES)
                                    INTERFACE_LINK_LIBRARIES "${BACKWARD_LIBRARIES}")
 endif()
 add_library(Backward::Backward ALIAS backward)
-
-###############################################################################
-# TESTS
-###############################################################################
-
-if(BACKWARD_TESTS)
-	enable_testing()
-
-	add_library(test_main OBJECT test/_test_main.cpp)
-
-	macro(backward_add_test src)
-		get_filename_component(name ${src} NAME_WE)
-		set(test_name "test_${name}")
-
-		add_executable(${test_name} ${src} ${ARGN} $<TARGET_OBJECTS:test_main>)
-
-		target_link_libraries(${test_name} PRIVATE Backward::Backward)
-
-		add_test(NAME ${name} COMMAND ${test_name})
-	endmacro()
-
-	# Tests without backward.cpp
-	set(TESTS
-		test
-		stacktrace
-		rectrace
-		select_signals
-		)
-
-	foreach(test ${TESTS})
-		backward_add_test(test/${test}.cpp)
-	endforeach()
-
-	# Tests with backward.cpp
-	set(TESTS
-		suicide
-		)
-
-	foreach(test ${TESTS})
-		backward_add_test(test/${test}.cpp ${BACKWARD_ENABLE})
-	endforeach()
-endif()
-if (BACKWARD_MPI)
-    add_executable(mpi_example mpi_example.cpp)
-    target_link_libraries(mpi_example PRIVATE Backward::Backward)
-endif()
-
-install(
-    FILES "backward.hpp"
-    DESTINATION ${CMAKE_INSTALL_INCLUDEDIR}
-)
-install(
-    FILES "BackwardConfig.cmake"
-    DESTINATION ${CMAKE_INSTALL_LIBDIR}/backward
-)
