diff --git a/kagen/CMakeLists.txt b/kagen/CMakeLists.txt
index dd90962..0c96f86 100644
--- a/kagen/CMakeLists.txt
+++ b/kagen/CMakeLists.txt
@@ -30,6 +30,3 @@ target_compile_options(kagen PRIVATE ${KAGEN_WARNING_FLAGS})
 
 # Allow library users to include kagen.h without kagen/ prefix
 target_include_directories(kagen INTERFACE ${CMAKE_CURRENT_SOURCE_DIR})
-
-install(TARGETS kagen DESTINATION lib)
-install(FILES kagen.h DESTINATION include)
